src/cmd/ksh93/edit/hexpand.c

```diff
@@ -590,7 +588,13 @@ int hist_expand(const char *ln, char **xp)
 				{
 					/* preset old with match from !?string? */
 					if(!sb.str[0] && wm)
-						sb.str[0] = sh_strdup(sfsetbuf(wm, (void*)1, 0));
+					{
+						char *sbuf = sfsetbuf(wm, (void*)1, 0);
+						int n = sftell(wm);
+						sb.str[0] = sh_malloc(n + 1);
+						sb.str[n] = '\0';
```

Comment:
This cannot be correct. Surely you mean `sb.str[0][n] = '\0';` as in the ksh2020 fix.